File: mobile/app/lib/features/chat/data/contact_display_name.dart
import 'user_contacts_repository.dart';

String buildContactDisplayName({String? firstName, String? lastName}) {
  final first = (firstName ?? '').trim();
  final last = (lastName ?? '').trim();
  return [first, last].where((x) => x.isNotEmpty).join(' ').trim();
}

String resolveContactDisplayName({
  required Map<String, ContactLocalProfile> contactProfiles,
  required String? contactUserId,
  required String fallbackName,
}) {
  final id = (contactUserId ?? '').trim();
  if (id.isEmpty) return fallbackName;
  final local = contactProfiles[id];
  if (local == null) return fallbackName;
  final display = (local.displayName ?? '').trim();
  if (display.isNotEmpty) return display;
  final composed = buildContactDisplayName(
    firstName: local.firstName,
    lastName: local.lastName,
  );
  return composed.isNotEmpty ? composed : fallbackName;
}

({String firstName, String lastName}) splitNameForContactForm(String source) {
  final normalized = source.replaceAll(RegExp(r'\s+'), ' ').trim();
  if (normalized.isEmpty) {
    return (firstName: '', lastName: '');
  }
  final parts = normalized.split(' ');
  if (parts.length == 1) {
    return (firstName: parts.first, lastName: '');
  }
  return (firstName: parts.first, lastName: parts.skip(1).join(' ').trim());
}

File: src/lib/contact-display-name.ts
import type { UserContactLocalProfile } from '@/lib/types';

export function buildContactDisplayName(input: {
  firstName?: string | null;
  lastName?: string | null;
}): string {
  const first = (input.firstName ?? '').trim();
  const last = (input.lastName ?? '').trim();
  return [first, last].filter(Boolean).join(' ').trim();
}

export function resolveContactDisplayName(
  contactProfiles: Record<string, UserContactLocalProfile> | null | undefined,
  contactUserId: string | null | undefined,
  fallbackName: string
): string {
  const id = (contactUserId ?? '').trim();
  if (!id || !contactProfiles) return fallbackName;
  const local = contactProfiles[id];
  if (!local) return fallbackName;
  const display = (local.displayName ?? '').trim();
  if (display) return display;
  const composed = buildContactDisplayName({
    firstName: local.firstName,
    lastName: local.lastName,
  });
  return composed || fallbackName;
}

export function splitNameForContactForm(sourceName: string): {
  firstName: string;
  lastName: string;
} {
  const normalized = sourceName.replace(/\s+/g, ' ').trim();
  if (!normalized) return { firstName: '', lastName: '' };
  const parts = normalized.split(' ');
  if (parts.length === 1) return { firstName: parts[0] ?? '', lastName: '' };
  return {
    firstName: parts.slice(0, 1).join(' ').trim(),
    lastName: parts.slice(1).join(' ').trim(),
  };
}
